File: vermyth/registry.py
from __future__ import annotations

from vermyth.schema import (
    ASPECT_CANONICAL_ORDER,
    Aspect,
    RegisteredAspect,
    _set_basis_version,
    current_basis_version,
    _reset_registered_aspects_for_tests,
    full_aspect_order,
    register_aspect,
    registered_aspects,
)


class AspectRegistry:
    _instance: AspectRegistry | None = None

    def __init__(self) -> None:
        self._canonical = list(ASPECT_CANONICAL_ORDER)
        self._by_name: dict[str, Aspect] = {a.name: a for a in ASPECT_CANONICAL_ORDER}
        for a in registered_aspects():
            self._by_name[a.name] = a

    @classmethod
    def get(cls) -> AspectRegistry:
        if cls._instance is None:
            cls._instance = AspectRegistry()
        return cls._instance

    @classmethod
    def reset(cls) -> None:
        cls._instance = None
        _reset_registered_aspects_for_tests()

    @property
    def dimensionality(self) -> int:
        return len(full_aspect_order())

    @property
    def full_order(self) -> tuple[Aspect, ...]:
        return full_aspect_order()

    def resolve(self, name: str) -> Aspect:
        return self._by_name[name]

    def is_registered(self, name: str) -> bool:
        v = self._by_name.get(name)
        return isinstance(v, RegisteredAspect)

    def registered_aspects(self) -> list[RegisteredAspect]:
        return registered_aspects()

    def register(self, aspect: RegisteredAspect) -> None:
        register_aspect(aspect)
        self._by_name[aspect.name] = aspect

    def current_basis_version(self) -> int:
        return current_basis_version()

    def set_basis_version(self, version: int) -> None:
        _set_basis_version(version)

File: typescript/src/registry.ts
import type { Aspect, RegisteredAspect } from "./schema/aspect.js";
import {
  ASPECT_CANONICAL_NAMES,
  aspectIdFromName,
  toRegisteredAspect,
  RegisteredAspectSchema,
} from "./schema/aspect.js";

let registeredAspects: RegisteredAspect[] = [];
let basisVersion = 0;

function readBasisVersionInternal(): number {
  return basisVersion;
}

function applyBasisVersionInternal(value: number): void {
  basisVersion = Math.max(0, Math.floor(value));
}

export function fullAspectOrder(): readonly Aspect[] {
  const canonical = ASPECT_CANONICAL_NAMES.map((n) => aspectIdFromName(n));
  return [...canonical, ...registeredAspects];
}

export function registeredAspectsList(): RegisteredAspect[] {
  return [...registeredAspects];
}

export function getBasisVersion(): number {
  return readBasisVersionInternal();
}

export function setBasisVersion(value: number): void {
  applyBasisVersionInternal(value);
}

export function resetRegisteredAspectsForTests(): void {
  registeredAspects = [];
  basisVersion = 0;
}

export function registerAspect(aspect: RegisteredAspect): void {
  for (const n of ASPECT_CANONICAL_NAMES) {
    if (n === aspect.name) {
      throw new Error(`Aspect ${JSON.stringify(aspect.name)} conflicts with canonical AspectID`);
    }
  }
  if (registeredAspects.some((a) => a.name === aspect.name)) {
    throw new Error(`Aspect ${JSON.stringify(aspect.name)} already exists`);
  }
  registeredAspects.push(aspect);
  basisVersion += 1;
}

export function parseRegisteredAspect(input: unknown): RegisteredAspect {
  return toRegisteredAspect(RegisteredAspectSchema.parse(input));
}

export class AspectRegistry {
  private static instance: AspectRegistry | undefined;

  static get(): AspectRegistry {
    if (!AspectRegistry.instance) {
      AspectRegistry.instance = new AspectRegistry();
    }
    return AspectRegistry.instance;
  }

  static reset(): void {
    AspectRegistry.instance = undefined;
    resetRegisteredAspectsForTests();
  }

  get dimensionality(): number {
    return fullAspectOrder().length;
  }

  get fullOrder(): readonly Aspect[] {
    return fullAspectOrder();
  }

  resolve(name: string): Aspect {
    try {
      return aspectIdFromName(name);
    } catch {
      const found = registeredAspects.find((a) => a.name === name);
      if (found) {
        return found;
      }
      throw new Error(`unknown aspect: ${JSON.stringify(name)}`);
    }
  }

  isRegistered(name: string): boolean {
    return registeredAspects.some((a) => a.name === name);
  }

  register(aspect: RegisteredAspect): void {
    registerAspect(aspect);
  }

  getBasisVersion(): number {
    return readBasisVersionInternal();
  }

  setBasisVersion(version: number): void {
    applyBasisVersionInternal(version);
  }
}
